src/data_structures.rs

```diff
@@ -71,17 +73,10 @@ pub struct ProverKey<
     pub index: Index<F>,
 }
 
-impl<F: PrimeField, PC: PolynomialCommitment<F>> Clone
-    for ProverKey<F, PC>
-where
-    PC::Commitment: Clone,
+impl<F: PrimeField, PC: PolynomialCommitment<F>> algebra::SemanticallyValid for ProverKey<F, PC>
 {
-    fn clone(&self) -> Self {
-        Self {
-            index_vk: self.index_vk.clone(),
-            index_comm_rands: self.index_comm_rands.clone(),
-            index: self.index.clone(),
-        }
+    fn is_valid(&self) -> bool {
+        self.index_vk.is_valid()
```

Comment:
The same here: we could add the checks that the arithmetizations of the R1CS matrices are correct, or at least do some lightweight checks as degree bounds, etc.